### PR TITLE
fix mTLS certificate check on agent to agent RPCs

### PR DESCRIPTION
**Pull request by lgfa29**
PR #11956 implemented a new mTLS RPC check to validate the role of the
certificate used in the request, but further testing revealed two flaws:

  1. client-only endpoints did not accept server certificates so the
     request would fail when forwarded from one server to another.
  2. the certificate was being checked after the request was forwarded,
     so the check would happen over the server certificate, not the
     actual source.

This PR checks for the desired mTLS level, where the client level
accepts both, a server or a client certificate. It also validates the
cercertificate before the request is forwarded.

#11956 was never released, so no CHANGELOG needed.

### CODE REVIEW

File: nomad/rpc_test.go [hunk 1]
```diff
@@ -1186,7 +1186,7 @@ func TestRPC_TLS_Enforcement_RPC(t *testing.T) {
 			name:   "local server/clients only rpc",
 			cn:     "server.global.nomad",
 			rpcs:   localClientsOnlyRPCs,
-			canRPC: false,
+			canRPC: true,
```
Comment (lgfa29) [inline nomad/rpc_test.go:1189]: This caused problem 1 to go unnoticed. Server certificates should be able to call client RPCs

File: nomad/rpc_test.go [hunk 2]
```diff
@@ -1274,18 +1274,22 @@ func TestRPC_TLS_Enforcement_RPC(t *testing.T) {
 			}
 
 			for method, arg := range tc.rpcs {
-				t.Run(fmt.Sprintf("nomad RPC: rpc=%s verify_hostname=true", method), func(t *testing.T) {
-					err := tlsHelper.nomadRPC(t, tlsHelper.mtlsServer, cfg, method, arg)
-
-					if tc.canRPC {
-						if err != nil {
-							require.NotContains(t, err, "certificate")
+				for _, srv := range []*Server{tlsHelper.mtlsServer1, tlsHelper.mtlsServer2} {
```
Comment (lgfa29) [inline nomad/rpc_test.go:1277]: This tests problem `2` by using a two server cluster and making requests to both. One of the requests will need to be forwarded.

**Review by schmichael — APPROVED** [review]
Great catch!

**Comment by lgfa29** [comment]
Hi @samed 

This PR hasn't been released yet so it can't be causing that error (the error message is also [slightly different](https://github.com/hashicorp/nomad/blob/main/nomad/rpc.go#L882) in `main`).

Did anything change in your mTLS setup? From the error message it seems like you may be using a wildcard certificate in your servers?

**Comment by lgfa29** [comment]
> but the upgrade was from 1.1.2 to 1.2.5

Ah OK, so this check was introduced in [1.1.4](https://github.com/hashicorp/nomad/issues/11084), that's why you weren't seeing it before.

**Pull request merged**